Compiler bug when copy embedded structure

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 21.04
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1) 10.3.0

getwd: /var/www/vproject
vmodules: /home/tanguy/.vmodules
vroot: /var/www/vproject/v
vexe: /var/www/vproject/v/v
vexe mtime: 2022-02-19 09:34:24
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 f8b8950.3fa6622

Git version: git version 2.30.2
Git vroot status: weekly.2022.07-48-g3fa6622e
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8-dirty

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/test/structs/struct_n_2_copy.v`
```v
module main

struct Animal{
    Duck 
    id int
}

struct Duck{
    RedDuck
    action string
}

struct RedDuck{
    color string
}

d := Animal{Duck{RedDuck{'red'}, 'fly'}, 1}
println(d)

e := Animal{...d}
println(e)

```

**What did you expect to see?**

Display new struct e.


**What did you see instead?**
```
==================
/tmp/v_1000/struct_n_2_copy.4624944979773238985.tmp.c:11347: error: field not found: RedDuck
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -230,13 +230,19 @@ fn (mut g Gen) struct_init(node ast.StructInit) {
 					g.write('.')
 				}
 				if node.is_update_embed {
-					embed_sym := g.table.sym(node.typ)
-					embed_name := embed_sym.embed_name()
-					g.write(embed_name)
-					if node.typ.is_ptr() {
-						g.write('->')
-					} else {
-						g.write('.')
+					update_sym := g.table.sym(node.update_expr_type)
+					_, embeds := g.table.find_field_from_embeds(update_sym, field.name) or {
+						ast.StructField{}, []ast.Type{}
+					}
+					for embed in embeds {
+						esym := g.table.sym(embed)
+						ename := esym.embed_name()
+						g.write(ename)
+						if embed.is_ptr() {
+							g.write('->')
+						} else {
+							g.write('.')
+						}
 					}
 				}
 				g.write(field.name)

--- a/vlib/v/tests/struct_init_with_multi_nested_embed_update_test.v
+++ b/vlib/v/tests/struct_init_with_multi_nested_embed_update_test.v
@@ -1,0 +1,28 @@
+module main
+
+struct Animal {
+	Duck
+	id int
+}
+
+struct Duck {
+	RedDuck
+	action string
+}
+
+struct RedDuck {
+	color string
+}
+
+fn test_struct_init_with_multi_nested_embed_update() {
+	d := Animal{Duck{RedDuck{'red'}, 'fly'}, 1}
+	println(d)
+
+	e := Animal{
+		...d
+		id: 2
+	}
+	println(e)
+	assert d.id == 1
+	assert e.id == 2
+}

